### ISSUE TITLE
Crash

### ISSUE THREAD

**Issue by bombez** [Donky-Network/DonkySDK-iOS-Modular#12]
Hi,
we are currently using the following library from cocoa pods 
Using Donky-CommonMessaging-Logic (4.8.3.1)
Using Donky-CommonMessaging-UI (4.8.3.1)
Using Donky-Core-SDK (4.8.3.1)
Using Donky-Push (4.8.3.1)
Using Donky-RichMessage-Inbox (4.8.3.1)
Using Donky-RichMessage-Logic (4.8.3.1)

we have followed the steps in the documentation, but I want to address a crash bug who appears when the app is in background and receive over 15 rich push notifications in a row, when a user re-open it after some time the app crashes . Could you please investigate?
Thanks.

**Comment by ben-moore** [comment]
Hi @bombez, can you confirm a few details please:
- app space 
- device details (model, OS version etc)
- user details for the device with the issue. 
- Is the issue repeatable? 
- do you have any crash logs or other diagnostics?

Thanks, 
Ben

**Comment by bombez** [comment]
Hi @ben-moore,

> device details 

currently I've test on iPhone 5s iOS 10 and iPhone 6s iOS 9 (so I think issue is reproducible on all devices type)

> user details for the device with the issue. 

what kind of informations do you need? 

> app space

what kind of informations do you need? (we are currently work for a third part company, so to share this kind of informations I need some permission)

> Is the issue repeatable?

 Yes, the issue is repeatable.

> do you have any crash logs or other diagnostics?

[logs.txt](https://github.com/Donky-Network/DonkySDK-iOS-Modular/files/645588/logs.txt)
I've attach the logs of the phone, I notice that: 
`INFO: Merging changes into main context: NSConcreteNotification 0x10f3256a0 {name = NSManagingContextDidSaveChangesNotification; object = <NSManagedObjectContext: 0x1701db5d0>; userInfo = {
    inserted = "{(\134n)}";
    managedObjectContext = "<_PFWeakReference: 0x170232bc0>";
    updated = "{(\134n    <DNNotification: 0x10d71e4d0> (entity: DNNotification; id: 0xd000000002480002 <x-coredata://4037BE5C-AFCD-45FA-96AF-4BBF67D958E2/DNNotification/p146> ; data: {\134n    acknowledgementDetails = \134"{\134\134n    result = delivered;\134\134n    sentTime = \134\134\134"2016-12-12T10:33:03+01:00\134\134\134";\134\134n    serverNotificationId = \134\134\134"c5240f6d-0993-4cfc-8702-c5341bfef25b\134\134\134";\134\134n    type = RichMessage;\134\134n}\134";\134n    audience = nil;\134n    content = nil;\134n    data = \134"{\134\134n    contextItems =     {\134\134n        MultipleRecipients = True;\134\134n        action = \134\134\134"quote_ready\134\134\134";\134\134n        \134\134\134"quote_number\134\134\134" = 1214;\134\134n        \134\134\134"user_id\134\134\134" = test16;\134\134n    };\134\134n    messageId = \134\134\134"3de46e20-1e99-45ea-8585-7eeb2d5`
or similar are printed many times so the app goes overflow

Thank you!

**Comment by donkybuild** [comment]
Hi @bombez, thanks for the info.  We'll review this and see if we can repro.

If you have any further information you can't share on GitHub, or if you have any other details related to support contracts etc then please email support@mobiledonky.com and reference this.

Thanks,
Ben

**Comment by ben-moore** [comment]
Hi @bombez, thanks for the info. We'll review this and see if we can repro.

If you have any further information you can't share on GitHub, or if you have any other details related to support contracts etc then please email support@mobiledonky.com and reference this.

Thanks,
Ben

**Comment by bombez** [comment]
Hi @ben-moore,
there's some updates on this?
are you be able to reproduce it?
Thanks

**Comment by bombez** [comment]
Hi,
I've found the problem, in the method 

`- (void)mergeChanges:(NSNotification *)notification {`

there are 2 selectors called by performOnMainThread that block all the operations because are called many times, I think are called recursively.

so I've tried to perform asynchronously 
```
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [mainContext performSelector:@selector(mergeChangesFromContextDidSaveNotification:) withObject:notification afterDelay:0];
        [mainContext performSelector:@selector(saveIfHasChanges:) withObject:notification afterDelay:0];
    });
```
and it works apparently.

Thanks

**Comment by ben-moore** [comment]
Hi @bombez.  We've reviewed your scenario in more detail and can indeed confirm an issue when processing a reasonable number of notifications during startup.  The resultant behaviour depends on the resources available on the device but it could well result in the crashes you have seen.  The root cause is a little deeper, we need to eliminate the iterative calls that trigger this scenario.  A fix will be included in the next SDK release.

**Comment by bombez** [comment]
Hi @ben-moore,
thanks for your work!!
If it's possible I would know the release date of fixed SDK so I can share it with client.
Thanks a lot for your support.

**Comment by illyabbi** [comment]
Hi @bombez, 
In a separate communication, @ben-moore mentioned last week a possible release end of this week.
@ben-moore, please correct if I'm mistaken?

Thank you both.

**Comment by ben-moore** [comment]
Hi @bombez and @illyabbi, the fix is in test, expected to complete today so you should expect a release before end of day tomorrow.

**Comment by ben-moore** [comment]
Fixed in release v4.8.4.1
